### ISSUE TITLE
Username/password auth when fetching from a binary cache?

### ISSUE THREAD

**Issue by adnelson** [NixOS/nix#950]
I want to set a machine to host pre-built binaries. However for security reasons, I want to put the server behind HTTPS with basic auth (username/password). This can be done by putting the password directly in the URL, but of course this is insecure. I have a patch to `download-from-binary-cache.pl` that looks for environment variables (`NIX_BINARY_CACHE_USERNAME` and `NIX_BINARY_CACHE_PASSWORD`) and sets the corresponding curl options when doing fetches (it also switches to using `WWW::Curl` instead of a curl `system` call, so that the password doesn't show up in a call to `ps aux`).

My patch is written against version 1.11.2, so I'm not sure how out-of-date it might be against master. In addition, I'm not sure if this is a feature that the nix maintainers would want to put in the platform, or if the environment variables I chose would be desirable. I thought I'd ask about this before working on submitting my modifications upstream.

**Comment by edolstra** [comment]
The binary cache substituter is gone in master, replaced by https://github.com/NixOS/nix/blob/master/src/libstore/http-binary-cache-store.cc, which uses libcurl.

**Comment by adnelson** [comment]
Oh, if it's gone why is `download-from-binary-cache.pl` still there?

Also looking at that file I don't see anything about username/password. Should I take a look at integrating that functionality?

**Comment by k0001** [comment]
@adnelson I'm looking forward to this this feature as well.

I see a problem with your environment variable approach, though: It doesn't support providing more than one username and password, whereas one could be needing more than one (it's my case, currently). I think the correct solution is to use the `netrc` functionality from curl. There could be an environment `NIX_CURL_NETRC` pointing to this file.

I'll try and implement this.

**Comment by copumpkin** [comment]
It might also be interesting to make it a bit more general, since it could make sense to do authenticated accesses against an S3 binary cache (the sort that isn't served up via CloudFront). I think Nix at least in principle can speak S3 directly nowadays for uploads, but I'm unsure whether it assumes direct HTTP for downloads. @edolstra?

**Comment by adnelson** [comment]
I've been using the patch I wrote back when I made this issue for a long time without issue; it works well enough but it certainly is primitive. I think that almost any responsible organization which wants to host their own binary cache would need some way to do auth, and I imagine that there would be several ways of authenticating (public key, basic auth, IAM, etc) which would be desirable.

That being said I'm very unfamiliar with the state of master.
